dynamic/discovery/discovery.go

```diff
@@ -125,14 +136,39 @@ func (rm *ResourceMap) refresh() {
 				apiResource.Version = gv.Version
 			}
 			gve.resources[apiResource.Name] = apiResource
-			// Remember how to map back from Kind to resource.
+			// Remember how to map back from Kind/subresource to resource.
 			// This is different from what RESTMapper provides because we already know
 			// the full GroupVersionKind and just need the resource name.
-			// Make sure we don't choose a subresource like "pods/status".
+			if strings.ContainsRune(apiResource.Name, '/') {
+				// Remember which resources are subresources, and map the kind to the main resource.
+				gve.subresources[apiResource.Name] = apiResource
+			} else {
+				// Make sure we don't choose a subresource like "pods/status".
+				gve.kinds[apiResource.Kind] = apiResource
+			}
+		}
+
+		// Group all subresources for a resource.
+		for i := range group.APIResources {
```

Comment:
You don't have to loop over all resources again. We already kept a map of only the ones that are subresources. In the second pass, you can just loop over all the subresources and insert them where they belong by looking up the main resource (if it exists). That would be O(n) instead of the current code which is O(n^2), and it also avoids allocating a new APIResource by reusing the one we already put in the map.

Reply:
Yes, your proposed way works better. This O(N^2) is to ensure this is something similar to what you intend for. I spent some time looking for O(N^1), but couldn't find any for recording all subresources for a resource, so I think O(N) is the best solution.